src/java/org/apache/cassandra/index/sai/disk/format/IndexDescriptor.java

```diff
@@ -0,0 +1,530 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.disk.format;
+
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Objects;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.google.common.io.Files;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.db.ClusteringComparator;
+import org.apache.cassandra.db.lifecycle.LifecycleNewTracker;
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.SSTableContext;
+import org.apache.cassandra.index.sai.StorageAttachedIndex;
+import org.apache.cassandra.index.sai.disk.PerIndexWriter;
+import org.apache.cassandra.index.sai.disk.PerSSTableWriter;
+import org.apache.cassandra.index.sai.disk.PrimaryKeyMap;
+import org.apache.cassandra.index.sai.disk.SearchableIndex;
+import org.apache.cassandra.index.sai.disk.io.IndexOutputWriter;
+import org.apache.cassandra.index.sai.memory.RowMapping;
+import org.apache.cassandra.index.sai.utils.IndexFileUtils;
+import org.apache.cassandra.index.sai.utils.PrimaryKey;
+import org.apache.cassandra.index.sai.utils.PrimaryKeyFactory;
+import org.apache.cassandra.io.sstable.Component;
+import org.apache.cassandra.io.sstable.Descriptor;
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.io.util.File;
+import org.apache.cassandra.io.util.FileHandle;
+import org.apache.cassandra.utils.FBUtilities;
+import org.apache.cassandra.utils.Pair;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.util.IOUtils;
+
+/**
+ * The {@link IndexDescriptor} is an analog of the SSTable {@link Descriptor} and provides version
+ * specific information about the on-disk state of a {@link StorageAttachedIndex}.
+ *
+ * The {@link IndexDescriptor} is primarily responsible for maintaining a view of the on-disk state
+ * of an index for a specific {@link org.apache.cassandra.io.sstable.SSTable}. It maintains mappings
+ * of the current on-disk components and files. It is responsible for opening files for use by
+ * writers and readers.
+ *
+ * It's remaining responsibility is to act as a proxy to the {@link OnDiskFormat} associated with the
+ * index {@link Version}.
+ */
+public class IndexDescriptor
+{
+    private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    public final Version version;
+    public final Descriptor descriptor;
+    public final ClusteringComparator clusteringComparator;
+    public final PrimaryKeyFactory primaryKeyFactory;
+    public final Set<IndexComponent> perSSTableComponents = Sets.newHashSet();
```

Comment:
So unlike `Descriptor`, about half of our state here is mutable. I guess I'd either verify that we're only supposed to use this from a single thread and throw a `@NotThreadSafe` on the class. Or make the changes we need to safely publish changes to these bookkeeping collections, etc. (Not looking at this super deeply ATM, but might get a chance to later...)

Reply:
I guess the fact that we "register" components on-demand in many methods below slightly complicates this.

Reply:
This is a valid point. The `IndexDescriptor` was designed to be immutable but the mutable elements have crept in. Realistically these mutable elements should be broken out into a separate on-disk utility class.

I am inclined to mark the class as `@NotThreadSafe` for now and raise a follow on ticket to break the class apart

WDYT?

Reply:
Let's leave this unresolved for now and come back to it. Still a ton of review to work through...

Reply:
EDIT: I'm having another look at this because it looks like this can be improved significantly by separating out the mutable and immutable state.

Reply:
So, having had a look at this, I have removed the mutable state from `IndexDescriptor` so the class is now immutable. Looking at the usages of the mutable state, it became clear that the usages were one-offs, such that there was little advantage in the loading of files into maps. Removing this state has made a lot of the methods far easier to understand so any minor degradation in performance from not caching file instances is gained by increased readability.